### ISSUE TITLE
Eventually / Retrying metaspace memory leak

### ISSUE THREAD

**Issue by francisdb** [scalatest/scalatest#1960]
sbt 1.4.5
AdoptOpenJDK Java 11.0.9.1
scalatest 3.2.3

SBT running tests repeatedly without forking. (`~test`)

Eventually is using Retrying
Retrying is an object that contains a scheduler
**This scheduler is never shut down** and recreated on every run (because of the sbt classloader)

We have a akka test using the akka testkit and Eventually

(getPreviousPeriodCountAndReset() returns a future)

```scala

class AkkaMetricsSpec
    extends TestKit(ActorSystem("MySpec"))
    with ImplicitSender
    with AsyncWordSpecLike
    with Matchers
    with BeforeAndAfterAll
    with Eventually {

  override def afterAll(): Unit =
    TestKit.shutdownActorSystem(system, 10.seconds, true)

  "The akka metrics" should {

    "correctly count" in {
      val metrics = new AkkaMetrics()(system)
      eventually(timeout(scaled(2.seconds)))(
        metrics
          .getPreviousPeriodCountAndReset()
          .map { count =>
            count mustEqual 19
          }
      )
    }
  }
}
```

This specific test causes metaspace out of memory errors, somehow the code inside eventually is captured and never released, even if the test succeeds successfully.
Every run leaves an extra thread pool-xxx-thread-1 (it would be better if these threads were at least named so that they are easier to track down.)
More and more classes get loaded an metaspace keeps growing on every run.

I will try to come up with a simple test project that reproduces this.

**Comment by cheeseng** [comment]
@francisdb Thanks for your report!

I did a bit into the code and I think the related scheduler is here: 

https://github.com/scalatest/scalatest/blob/3.2.x-new/jvm/core/src/main/scala/org/scalatest/enablers/Retrying.scala#L59

It is a lazy val in object Retrying that intended to be shared, my theory is that every test run in sbt will create a new test class loader instance which reloads the object Retrying class thus its scheduler instance, without calling shutdown in previous instance.  I am thinking if the better approach is to create new scheduler instance at this line: 

https://github.com/scalatest/scalatest/blob/3.2.x-new/jvm/core/src/main/scala/org/scalatest/enablers/Retrying.scala#L114

and call shutdown() right after the task is submitted.

**Comment by francisdb** [comment]
I still wonder why the lambda that captures the assertion is not released after the eventually finished

**Comment by cheeseng** [comment]
@francisdb May I know how do you check if the lambda is not released?  I can reproduce the extra thread problem you mentioned, and submitted the following PR for it: 

https://github.com/scalatest/scalatest/pull/1962

Cheers.

**Comment by francisdb** [comment]
I use visualvm and monitor the meta space / loaded and unloaded classes. Forcing GC from time to time. All this while repeating this single test case in an open sbt session.

With the example code above the meta space keeps growing and more and more classes are loaded. (It keeps my code and the  the actorsystem from previous runs loaded )

**Comment by francisdb** [comment]
Do you want me to come up with a small repro project?

**Comment by cheeseng** [comment]
@francisdb A small repro project will definitely be helpful.  :)

**Comment by francisdb** [comment]
@cheeseng the reproducer is here: https://github.com/francisdb/scalatestleak

The readme explains how to use it

**Comment by cheeseng** [comment]
@francisdb Thanks for the reproducer, I tried it out today with the new code it seems to be ok, you can try my branch here: 

https://github.com/cheeseng/scalatest/tree/feature-eventually-test

I have added your reproducer as eventually-test project, so after you checkout the branch you may do: 

```
> sbt
root> project eventually-test
eventually-test> test
```

It seems to work fine on my machine on graalvm ce 21.0.0 (jdk version 11.0.10), can you cross check if the latest code fix the problem you're seeing?

Cheers.

**Comment by francisdb** [comment]
```
 cheeseng_scalatest git:(master) sbt
[info] Loading project definition from /.../cheeseng_scalatest/project/project
java.lang.NullPointerException
	at java.base/java.util.regex.Matcher.getTextLength(Matcher.java:1770)
	at java.base/java.util.regex.Matcher.reset(Matcher.java:416)
	at java.base/java.util.regex.Matcher.<init>(Matcher.java:253)
	at java.base/java.util.regex.Pattern.matcher(Pattern.java:1133)
	at java.base/java.util.regex.Pattern.split(Pattern.java:1261)
	at java.base/java.util.regex.Pattern.split(Pattern.java:1334)
	at sbt.IO$.pathSplit(IO.scala:744)
	at sbt.IO$.parseClasspath(IO.scala:859)
	at sbt.compiler.CompilerArguments.extClasspath(CompilerArguments.scala:62)
	at sbt.compiler.AggressiveCompile.withBootclasspath(AggressiveCompile.scala:50)
	at sbt.compiler.AggressiveCompile.compile2(AggressiveCompile.scala:83)
	at sbt.compiler.AggressiveCompile.compile1(AggressiveCompile.scala:70)
	at sbt.compiler.AggressiveCompile.apply(AggressiveCompile.scala:45)
	at sbt.Compiler$.apply(Compiler.scala:74)
	at sbt.Compiler$.apply(Compiler.scala:65)
	at sbt.Defaults$.sbt$Defaults$$compileTaskImpl(Defaults.scala:789)
	at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:781)
	at sbt.Defaults$$anonfun$compileTask$1.apply(Defaults.scala:781)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:40)
	at sbt.std.Transform$$anon$4.work(System.scala:63)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:226)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:17)
	at sbt.Execute.work(Execute.scala:235)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:226)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:159)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:28)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by cheeseng** [comment]
@francisdb Don't run in master branch, go to branch feature-eventually-test: 

> git checkout feature-eventually-test

**Comment by francisdb** [comment]
> @francisdb Don't run in master branch, go to branch feature-eventually-test:
> 
> > git checkout feature-eventually-test

Oh my, It was early in the morning I guess 😴

I can confirm this is fixed, tried on
```
JVM: OpenJDK 64-Bit Server VM (11.0.9.1+1, mixed mode)
Java: version 11.0.9.1, vendor AdoptOpenJDK
```

**Comment by cheeseng** [comment]
@francisdb Thanks for trying and confirming the fix!  :)

Hopefully this can be included in the upcoming 3.2.4 asap.

**Comment by francisdb** [comment]
That would be great, thanks again for the quick responses

**Comment by bvenners** [comment]
This is merged and will be released in 3.2.5. Closing.
